File: mcs/class/Mono.Ext/test1.c
#include <stdio.h>
#include <stdlib.h>
#include <assert.h>

typedef unsigned UINT, UINT32;

#ifdef _MSC_VER
typedef unsigned __int64 UINT64;
typedef __int64 INT64;
#define I64 "I64"
#define DLLEXPORT
#else
typedef unsigned long long UINT64;
typedef long long INT64;
#define I64 "ll"
#define DLLEXPORT __declspec(dllexport) // for debugging
#define _strtoui64 strtoull
#endif

#if _MSC_VER

UINT64 _rotl64(UINT64 value, int count);
#pragma intrinsic(_rotl64)

UINT64 _rotr64(UINT64 value, int count);
#pragma intrinsic(_rotr64)

UINT32 _rotl(UINT32 value, int count);
#pragma intrinsic(_rotl)

UINT32 _rotr(UINT32 value, int count);
#pragma intrinsic(_rotr)

DLLEXPORT UINT64 intrinsic_rotate_left_64(UINT64 value, int count) { return _rotl64(value, count); }
DLLEXPORT UINT64 intrinsic_rotate_right_64(UINT64 value, int count) { return _rotr64(value, count); }
DLLEXPORT UINT32 intrinsic_rotate_left_32(UINT32 value, int count) { return _rotl(value, count); }
DLLEXPORT UINT32 intrinsic_rotate_right_32(UINT32 value, int count) { return _rotr(value, count); }

#define MSC_ASSERT(x) assert(x)

#else
#define MSC_ASSERT(x)
#endif

#define ROTATE(type, name, direction, other_direction)		\
DLLEXPORT type name(type value, int count)					\
{															\
	type const result = (value direction count) | (value other_direction (sizeof(type) * 8 - count));	\
	MSC_ASSERT(result == intrinsic_##name(value, count));	\
	return result;											\
}

ROTATE(UINT32, rotate_left_32, <<, >>)
ROTATE(UINT32, rotate_right_32, >>, <<)
ROTATE(UINT64, rotate_left_64, <<, >>)
ROTATE(UINT64, rotate_right_64, >>, <<)

void test(UINT64 value, int count)
{
	printf("%X rol32 %d = %X\n", (int)value, count, rotate_left_32((UINT32)value, count));
	printf("%X ror32 %d = %X\n", (int)value, count, rotate_right_32((UINT32)value, count));
	printf("%" I64 "X rol64 %d = %" I64 "X\n", value, count, rotate_left_64(value, count));
	printf("%" I64 "X ror64 %d = %" I64 "X\n", value, count, rotate_right_64(value, count));
}

int main(int argc, char** argv)
{
	if (argc == 3)
	{
		test(_strtoui64(argv[1], 0, 0), _strtoui64(argv[2], 0, 0));
		return 0;
	}
	for (int count = 1; count <= 16; ++count)
	{
		test(-1, count);
	}
	for (UINT64 value = 0; value <= 16; ++value)
	{
		for (int count = 1; count <= 16; ++count)
		{
			test(value ? value : -1, count);
		}
	}
	return 0;
}

File: mcs/class/Mono.Ext/test1.cs
import test1data;
public class test1
{
   static uint RotateLeft(uint value, int count)
   {
        return (value << count) | (value >> (32 - count));
    }

  
    static uint RotateRight(uint value, int count)
    {
        return (value >> count) | (value << (32 - count));
    }


    static ulong RotateLeft(ulong value, int count)
    {
        return (value << count) | (value >> (64 - count));
    }

    static ulong RotateRight(ulong value, int count)
    {
        return (value >> count) | (value << (64 - count));
    }

    void Assert(bool value, string name)
    {
        if (value) return;
        System.Console.WriteLine("assertion failure:{0}", name);
        MonoExt.DebugBreak();
    }

    static void test1(ulong value, int count)
    {
        uint c = 0;
        uint d = 0;
        ulong e = 0;
        ulong f = 0;
        uint g = 0;
        uint h = 0;
        ulong i = 0;
        ulong j = 0;

        MonoExt.DebugBreak();
        double t1 = (double)MonoExt.ReadTimeStampCounter();

        int ii;
        int iterations = 1000 * 10;

        for (ii = 0; ii < iterations; ++ii)
        {
            c = RotateLeft((uint)value, count);
            d = RotateRight((uint)value, count);
            e = RotateLeft((ulong)value, count);
            f = RotateRight((ulong)value, count);
        }

        double t2 = (double)MonoExt.ReadTimeStampCounter();

        for (ii = 0; ii < iterations; ++ii)
        {
            g = MonoExt.RotateLeft((uint)value, count);
            h = MonoExt.RotateRight((uint)value, count);
            i = MonoExt.RotateLeft((ulong)value, count);
            j = MonoExt.RotateRight((ulong)value, count);
        }

        ulong t3 = MonoExt.ReadTimeStampCounter();

        double slow = t2 - t1;
        double fast = t3 - t2;

        System.Console.WriteLine("{0} {1} {2}", slow, fast, (slow - fast) / slow);

        System.Console.WriteLine("{0:X} rol32 {1:X} = {2:X}", value, count, c);
        System.Console.WriteLine("{0:X} ror32 {1:X} = {2:X}", value, count, d);
        System.Console.WriteLine("{0:X} rol64 {1:X} = {2:X}", value, count, e);
        System.Console.WriteLine("{0:X} ror64 {1:X} = {2:X}", value, count, f);

        System.Console.WriteLine("{0:X} rol32 {1:X} = {2:X}", value, count, g);
        System.Console.WriteLine("{0:X} ror32 {1:X} = {2:X}", value, count, h);
        System.Console.WriteLine("{0:X} rol64 {1:X} = {2:X}", value, count, i);
        System.Console.WriteLine("{0:X} ror64 {1:X} = {2:X}", value, count, j);
    }


    static void Main(string[] args)
    {
        System.Console.WriteLine("Hello World!");
 
        if (args.Length > 1)
        {
            ulong value = ulong.Parse(args[0]);
            int count = int.Parse(args[1]);
            test(value, count);
            return;
        }
        for (ulong value = 1; value <= 4; ++value)
        {
            for (int count = 1; count <= 4; ++count)
            {
                test(value, count);
            }
        }
        for (long value = -1; value <= 16; ++value)
        {
            for (int count = 0; count <= 16; ++count)
            {
            //  test(value, count);
            }
        }
    }
}
